saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/assertion/SAMLAttributeValueParser.java

```diff
@@ -141,4 +143,27 @@ public static String parseAnyTypeAsString(XMLEventReader xmlEventReader) throws
         }
     }
 
+    private static String addNamespaceWhenMissing(Deque<String> definedPrefixes, XMLEventWriter writer, StartElement startElement)
+            throws XMLStreamException {
+
+        if (startElement.getName().getPrefix() != null && !startElement.getName().getPrefix().isEmpty()
+                && !definedPrefixes.contains(startElement.getName().getPrefix())) {
+
+            Iterator<Namespace> namespaces = startElement.getNamespaces();
+            boolean hasNamespace = false;
+            while (namespaces.hasNext()) {
+                Namespace namespace = namespaces.next();
+                if (namespace.getNamespaceURI().equals(startElement.getName().getNamespaceURI())) {
+                    hasNamespace = true;
+                    break;
+                }
+            }
+            if (!hasNamespace) {
+                XMLEventFactory xmlEventFactory = XMLEventFactory.newInstance();
+                writer.add(xmlEventFactory.createNamespace(startElement.getName().getPrefix(), startElement.getName().getNamespaceURI()));
+                return startElement.getName().getPrefix();
```

Comment:
FTR: I agree that we need not consider extreme cases where the same prefix would (per context) map to two different namespaces like in here:
```xml
<a:a xmlns:a="uri:1">
   <a:b xmlns:a="uri:2"/>
</a:a>
```
Still I'm wondering whether to cover this case:
```xml
<a xmlns="uri:1">
  <b xmlns="uri:2"/>
</a>
```
But I guess we can leave this only if the need arises too.

Reply:
I agree different namespaces with the same prefixes don't make sense. Missing prefixes are working (and are tested now).